### ISSUE TITLE
Set workspace/monitor default master layout orientation

### ISSUE THREAD

**Issue by robertoash** [hyprwm/Hyprland#3174]
### Description

I may have missed something but I have scoured the wiki for a solution to this and have been unable to find one. 

I use 2 monitors, one horizontal and one vertical. I would like to set the default orientation of the master layout to "top" on the vertical monitor and to "left" on the horizontal one.

I have added keybinds as a workaround. The issue is that I open several apps on startup and they all open with the wrong layout on my vertical monitor. I have to manually go into every workspace on that monitor and change the layout with the keybinds.

These are the keybinds I've set up:
```
bind = SUPER CTRL SHIFT, up, layoutmsg, orientationleft
bind = SUPER CTRL SHIFT, right, layoutmsg, orientationleft
bind = SUPER CTRL SHIFT, left, layoutmsg, orientationtop
bind = SUPER CTRL SHIFT, down, layoutmsg, orientationtop
```
These let me "toggle" between the two orientations. I have not found a way to use the layoutmsg dispatcher together with `exec-once`.

This is an example of two apps I open on startup:
```
exec-once = [workspace 1 silent] vscodium
exec-once = [workspace 1 silent] slack
```
I have not been able either to find a way to tell Hyprland to change the orientation to bottom before or after launching slack on workspace 1, for example.

Another solution (even nicer in my opinion) would be to be able to set a config so that all workspaces that spawn on a specific monitor are set to default to a specific layout. I have bound workspaces to the vertical monitor so having this config for the monitor would achieve the desired results.

**TLDR:** layoutmsg can be used to change the orientation of the master layout but the wiki only mentions how it can be used together with keybinds and does not mention how or if it can be used together with `exec-once`, a workspace rule or a monitor config.

Any ideas? If not, then maybe it's a good feature to request.

**Comment by vaxerski** [comment]
script?

listen to the activeworkspace event and issue a layoutmsg based on that

**Comment by robertoash** [comment]
It would be nice if the functionality was built in in the same way that you can easily do a transform on a monitor to make it vertical. 

In any case, a script could be a workaround so, after a lot of trial and error, I came up with this:

```
#! /bin/bash
# This script is used to listen for newly created workspaces using hyprctl
# and changing the layout to orientationtop if the workspace is on a vertical monitor

#!/bin/bash

# Variable to store workspace information from the previous run
previous_run_workspaces=""

# Delay to give time for the workspaces to be created on boot
sleep 5

####################################################
############## Get a list of vertical monitors ################
####################################################

# Get a JSON of all monitors using hyprctl monitors -j
ALL_MONITORS=$(hyprctl monitors -j)
# Filter out only the monitors that have a transform property of 1
VERTICAL_MONITORS=$(echo $ALL_MONITORS | jq '.[] | select(.transform == 1)')
# Get the names of vertical monitors
VERTICAL_MONITOR_NAMES=$(echo $VERTICAL_MONITORS | jq -r '.name')

####################################################
### Apply the vertical layout to all workspaces on vertical monitors ###
####################################################

while true; do
    # Get the workspace JSON for the current run
    current_run_workspaces=$(hyprctl workspaces -j | jq -r '.[] | select(.name != "special") | "\(.monitor):\(.id)"')
    
    # Compare with the previous run to identify new workspaces
    new_workspaces=$(comm -13 <(echo "$previous_run_workspaces" | tr ' ' '\n' | sort) <(echo "$current_run_workspaces" | tr ' ' '\n' | sort))
    
    # Update the variable with current workspaces for the next run
    previous_run_workspaces="$current_run_workspaces"
    
    # Iterate through each vertical monitor
    for monitor in $VERTICAL_MONITOR_NAMES; do
        # Filter the new workspaces for the current monitor
        monitor_new_workspaces=$(echo "$new_workspaces" | grep "^$monitor:" | cut -d: -f2-)
        
        # Cycle through each workspace and set the layout to vertical if not processed
        while read -r workspace_id; do
            # Check if workspace_id is not empty
            if [ -n "$workspace_id" ]; then
                # Select the workspace
                hyprctl dispatch "workspace $workspace_id" > /dev/null  # Redirect "ok" to /dev/null
                # Set the layout to vertical
                hyprctl dispatch "layoutmsg orientationtop" > /dev/null  # Redirect "ok" to /dev/null
                #echo "Orientationtop set for workspace $workspace_id on monitor $monitor"
            fi
        done <<< "$monitor_new_workspaces"
    done
    
    sleep 5
done

```

**Comment by cjber** [comment]
Taking inspiration from @robertoash and @ejiek, I have my own solution using IPC. Only works with one vertical monitor (which is all I have), given I just check the `$VERTICAL_MONITOR_NAMES `to make sure I execute the command on the correct monitor.

```bash
#!/bin/sh

# Get a JSON of all monitors using hyprctl monitors -j
ALL_MONITORS=$(hyprctl monitors -j)
# Filter out only the monitors that have a transform property of 1
VERTICAL_MONITORS=$(echo $ALL_MONITORS | jq '.[] | select(.transform == 1)')
# Get the names of vertical monitors
VERTICAL_MONITOR_NAMES=$(echo $VERTICAL_MONITORS | jq -r '.name')

handle() {
    case $1 in
        focusedmon*)
            if [[ $line == "focusedmon>>$VERTICAL_MONITOR_NAMES"* ]]; then
                hyprctl dispatch "layoutmsg orientationtop" > /dev/null  # Redirect "ok" to /dev/null
            else
                hyprctl dispatch "layoutmsg orientationcenter" > /dev/null  # Redirect "ok" to /dev/null
            fi ;;
    esac
}

socat -U - UNIX-CONNECT:/tmp/hypr/$HYPRLAND_INSTANCE_SIGNATURE/.socket2.sock | while read -r line; do handle "$line"; done
```

There may be a more elegant solution. The biggest issue currently is that the orientation doesn't change the first time the monitor is focussed if there are no open windows.

Used [https://wiki.hyprland.org/hyprland-wiki/pages/IPC/](url) for help.

**Comment by ejiek** [comment]
@ciber thanks for sharing your solution!

I'm somewhat happy to see that I'm not the only one suffering from an empty workspace ignoring my attempts to set its orientation.

[hyprland-auto-layout](https://github.com/ejiek/hyprland-auto-layout) is my playground to tackle this problem. Unfortunately, it's not ready to be used yet, but I'm getting it there.
Right now I've implemented two approaches:

**Fire Once** - that goes through all the workspaces as sets their orientation and then exits. This one is only possible when there are static bindings of workspaces to particular monitor. Furthermore it relies on parsing `hyprland.conf` to figure out a list of available workspaces.

**Daemonized** - listens for a workspace change through IPC. Runs is a background.

There are two imprtant assumptions:

- Orientation is based on a `transformation` parameter of a monitor.
- It needs at least one window to be present in a workspace to set its orientation. So it starts one if the ws is empty. It's a workaround, we need a proper fix in hyprland itself.

**Comment by thejch** [comment]
Worth noting that this is actually now possible, with the following workspace rule, for example:
`workspace = m[HDMI-A-1], layoutopt:orientation:top`
All workspaces created on monitor "HDMI-A-1" will be in the "top" orientation

**Comment by khalid151** [comment]
> Worth noting that this is actually now possible, with the following workspace rule, for example: `workspace = m[HDMI-A-1], layoutopt:orientation:top` All workspaces created on monitor "HDMI-A-1" will be in the "top" orientation

How to override the layout option? I mean, after I set such rule, I can no longer cycle between orientations and I'd be stuck with the top orientation. This is less than ideal, as I have to change it sometimes.
